numpyro/distributions/continuous.py

```diff
@@ -290,6 +291,112 @@ def infer_shapes(concentration):
 
 
+class EulerMaruyama(Distribution):
+    arg_constraints = {"t": constraints.ordered_vector}
+
+    def __init__(self, t, sde_fn, init_dist, validate_args=None):
+        self.t = t
+        self.dt = jnp.diff(t)
+        self.sde_fn = sde_fn
+        self.init_dist = init_dist
+
+        if init_dist.batch_shape:
+            raise ValueError(
+                "Init distribution is expected to have no batch dimension."
+            )
+        if jnp.ndim(t) != 1:
+            raise ValueError("`t` is expected to have 1 dimension.")
+        event_shape = t.shape + init_dist.event_shape
+        batch_shape = ()
+
+        super(EulerMaruyama, self).__init__(
+            batch_shape, event_shape, validate_args=validate_args
+        )
+
+    @constraints.dependent_property
+    def support(self):
+        return constraints.independent(constraints.real, self.event_dim)
+
+    def sample(self, key, sample_shape=()):
+        assert is_prng_key(key)
+        sample_size = np.prod(sample_shape)
+        sample_shape_ = (sample_size,) if sample_shape else ()
+
+        def step(y_curr, xs):
+            noise_curr, t_curr, dt_curr = xs
+            f, g = self.sde_fn(y_curr, t_curr)
+            mu = y_curr + dt_curr * f
+            sigma = jnp.sqrt(dt_curr) * g
+            y_next = mu + sigma * noise_curr
+            return y_next, y_next
+
+        noises = random.normal(
+            key, shape=sample_shape_ + (self.event_shape[0] - 1,) + self.event_shape[1:]
+        )
+        inits = self.init_dist.sample(key, sample_shape=sample_shape_)
+
+        def scan_fn(init, noise):
+            return scan(step, init, (noise, self.t[:-1], self.dt))
+
+        if sample_shape:
+            _, sde_out = vmap(scan_fn)(inits, noises)
+            sde_out = jnp.concatenate([inits[:, None], sde_out], axis=1)
+            sde_out = jnp.reshape(sde_out, sample_shape + self.event_shape)
+        else:
+            _, sde_out = scan_fn(inits, noises)
+            sde_out = jnp.concatenate([inits[None], sde_out], axis=0)
+
+        return sde_out
+
+    @validate_sample
+    def log_prob(self, value):
+        sample_shape = lax.broadcast_shapes(
+            value.shape[: -self.event_dim], self.batch_shape
+        )
+        value = jnp.broadcast_to(value, sample_shape + self.event_shape)
+
+        if sample_shape:
+            reshaped_value = value.reshape((-1,) + self.event_shape)
+            xtm1 = reshaped_value[:, :-1]
+            xt = reshaped_value[:, 1:]
+            value0 = reshaped_value[:, 0]
+
+            f, g = vmap(vmap(self.sde_fn), (0, None))(xtm1, self.t[:-1])
+
+            f = f.reshape(sample_shape + f.shape[1:] + (1,) * (xt.ndim - f.ndim))
+            g = g.reshape(sample_shape + g.shape[1:] + (1,) * (xt.ndim - g.ndim))
+            xtm1 = xtm1.reshape(sample_shape + xtm1.shape[1:])
+            xt = xt.reshape(sample_shape + xt.shape[1:])
+            value0 = value0.reshape(sample_shape + value0.shape[1:])
```

Comment:
Nit: Can we reshape the final log_prob to sample_shape instead of here?

Reply:
Actually, this way is more verbose but easier to follow. Thanks! Pls ignore my comment.

Reply:
I guess we can move the task of adding missing event dimensions of f and g outside of this if/else. Something like
```
# add missing event dimensions
batch_dim = len(sample_shape)
f = f.reshape...
g = g.reshape...
```

Reply:
move these codes after if/else.